README
    Copyright (c) 1996 Open Market, Inc.
    See the file "LICENSE.TERMS" for information on usage and redistribution
    of this file, and for a DISCLAIMER OF ALL WARRANTIES.

    Copyright (c) 1996 Sven Verdoolaege
    No additional restrictions/warranties.

This is a Fast CGI module for perl. It's based on the FCGI module
that comes with Open Market's FastCGI Developer's Kit, but does
not require you to recompile perl.
It does require that perl is compiled with sfio though.
Therefore you'll need at least perl 5.003_02 and you'll have
to have configured it with eg './Configure -Duseperlio -Dusesfio' .

See http://www.fastcgi.com/ for more information about fastcgi.
Lincoln D. Stein's perl CGI module also contains some information
about fastcgi programming.

See echo.fpl for an example on how to use this module.

To install, do the usual

perl Makefile.PL
make
make install

If Makefile.PL finds an already installed (exported) fcgi library
it will use that instead of the included files. Edit Makefile.PL
if you don't want that or if it doesn't find them.

Sven Verdoolaege
skimo@dns.ufsia.ac.be


ChangeLog
Release history for FCGI

Version 0.77 --  2014-08-05

  o re-release to remove build artifacts that should not have been shipped

Version 0.76 --  2014-08-05

  o On Android, set TMPDIR before calling configure (RT#97680, Brian Fraser)

Version 0.75 --  2014-07-17

   o deprecated APIs removed (chansen)
   o broken PP implementation removed (chansen)
   o retooled distribution so FCGI.pm and FCGI.xs exist as-is, rather than
     being generated by FCGI.PL and FCGI.XL (chansen)

Version 0.74 --  24 Sep 2011 <rafl@debian.org>
   o Stop leaking information across requests when using the deprecated
     and undocumented old FCGI interface. This is CVE-2011-2766.
   o Only discard input stream if FCGI_KEEP_CONN is set in
     FCGI_BeginRequestBody flags.

Version 0.73 --  19 May 2011 <rafl@debian.org>
   o Stop claiming we ship a file called -e in the MANIFEST.

Version 0.72 --  19 May 2011 <rafl@debian.org>
   o Clean up Makefile.PL and restore compatibility with recent
     ExtUtils::MakeMaker versions.

Version 0.71_03 --  28 Apr 2011 <bobtfish@bobtfish.net>
   o Remove support for sfio which is an optional (and not enabled by default)
     compile option to perl that is never used.
   o Fix FCGI::Stream::READ() to warn() instead of croak() incase of
     wide characters which cannot be gracefully downgraded.
   o Fix warnings due to wide characters being mangled to note that accepting
     them is deprecated and will stop working at some point.
   o Various fixes to FCGI::Stream::READ() to improve handling of error and
     edge cases.
     - croak if called with invalid number of arguments
     - croak if length is negative
     - croak if offset is outside string
     - pad scalar if offset is greater than length
   o Fix in FCGX_Finish_r to discard any remaining data in input stream
     which otherwise ends up in next request. This fixes multiple requests
     being broken if something goes wrong whilst reading the initial request.
     This discarding is done silently, as RFC 3875 says a script is not
     obliged to read any of the data.
   o Fixed indent style and braces to be consistent, swapped tabs for spaces
     in indenting.

Version 0.71_02 --  28 Apr 2011 <bobtfish@bobtfish.net>
   o Change the Request function to pass FAIL_ON_INTR into the XS
     RequestX function. This prevents the fcgi C client code from
     looping around their accept() call. This change means that
     when using CGI::Fast, and the process recieves SIGTERM or
     SIGHUP, the error statusis correctly passed back up, allowing process
     managers (such as FCGI::ProcManager) to correctly handle cleanly exiting.

Version 0.71_01 --   24 Aug 2010 <bobtfish@bobtfish.net>
   o Restore old behavior when un-downgradeable uft8 is sent to FCGI.
     The first time this happens, a warning will be issued, but subsequently
     the bytes will be sent through raw (causing double encoding etc).
     If the character string can be downgraded safely, then it will still be.
     use warnings FATAL => 'utf8'; can be used to make undowngradeable strings
     throw an exception.
   o Fix PRINT retval (Closes: RT#57697).

Version 0.71 --      1 Apr 2010  <rafl@debian.org> Florian Ragwitz
    o Fix some more defined(%hash) warnings on perl 5.12.

Version 0.70 --     22 Mar 2010  <bobtfish@bobtfish.net> Tomas Doran
    o Fix use of defined %hash which becomes deprecated in perl 5.12

Version 0.69 --     15 Feb 2010  <mst@shadowcat.co.uk> Matt S Trout
    o No changes since the previous development release.

Version 0.68_02 --   13 Jan 2010  <mst@shadowcat.co.uk> Matt S Trout
    o Make the PRINT method return a boolean value rather than the
	  number of bytes written, previous patch was incorrect.

Version 0.68_01 --   10 Jan 2010  <mst@shadowcat.co.uk> Matt S Trout
    o Force signal handler installation so that we correctly install handlers
      for SIGPIPE. Fixes RT#5100 <bobtfish@bobtfish.net>
    o Make the PRINT method return the number of bytes written rather than
      undef to be consistent with the IO:: interface. Fixes RT#24347
      <David Dick>
    o Fix UTF-8 double encoding when FCGI is passed octets by downgrading
      them into bytes correctly. Fixes RT#52400 <chansen@cpan.org>

Version 0.68 --     31 Dec 2009  <mst@shadowcat.co.uk> Matt S Trout
    o No changes since the previous development release.

Version 0.67_01 --   20 Dec 2009  <mst@shadowcat.co.uk> Matt S Trout
    o Add FILENO method which returns a defined but invalid value to
      placate things such as IPC::Run which call fileno to check if a
      filehandle is open.
      Closes bugs:
      http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=544540
      http://rt.cpan.org/Public/Bug/Display.html?id=50972
      Removes need for upstream patch in OpenBSD <bobtfish@bobtfish.net>
    o Call the fcgi lib's attach/detach <robs@fastcgi.com>

Version 0.67 --    22 December 2002   <skimo@kotnet.org> Sven Verdoolaege

    o Fixes for pure perl version based on report and patch
      from "Kurtis D. Rader" <kdrader@us.ibm.com>
    o FCGI_UndoBinding perl 5.8.0 compatibility
      Reported by Marko Asplund <aspa@kronodoc.fi>
    o Fix problem with fcgi_config.h on win32.
      Reported by Igor Franchuk <softdep@rol.ru>
    o Add minimal tests

Version 0.66 --    5 September 2002   <skimo@kotnet.org> Sven Verdoolaege

    o perl 5.8.0 compatibility fix by Autrijus
    o library fixes from Rob

Version 0.65 --    19 February 2002   <skimo@kotnet.org> Sven Verdoolaege

    o fix perl 5.005 compatibility problem
    o fix strict warning

Version 0.64 --    25 September 2001   <skimo@kotnet.org> Sven Verdoolaege
Version 0.63 --    24 September 2001   <skimo@kotnet.org> Sven Verdoolaege

    o Update build process

Version 0.62 --    21 September 2001   <skimo@kotnet.org> Sven Verdoolaege

    o Move version number to separate file

Version 0.61 --    20 September 2001   <skimo@kotnet.org> Sven Verdoolaege

    o Fix refcounting bug

Makefile.PL
# $Id: Makefile.PL,v 1.33 2002/12/15 19:40:19 skimo Exp $

use 5.006;
use ExtUtils::MakeMaker;
use IO::File;
use Config;
use Cwd 'cwd';
use Getopt::Long;
use File::Copy qw(copy);

@h1 = qw(fastcgi.h fcgiapp.h fcgimisc.h fcgios.h);
@h = (@h1, 'fcgi_config.h');
@o = qw(FCGI.o);
@dist1 = qw(LICENSE.TERMS);
@dist2 = qw(fcgiapp.c os_unix.c os_win32.c);
@dist3 = (@h1, qw(fcgi_config_x86.h));

GetOptions ("use-installed:s" => \$useinstalled);

$libfound = 0;
@libs = ();

my $cwd = cwd();
my $devkit = "$cwd/..";

if (defined $useinstalled) {
    require ExtUtils::Liblist;
    my $libspec = $useinstalled ? "-L$useinstalled/lib " : "";
    $libspec .= "-lfcgi";
    my @l = MM->ext($libspec);
    if ($l[0] || $l[1] || $l[2]) {
        $prefix = "$useinstalled/include" if $useinstalled;
        $libfound = 1;
        push @libs, $libspec;
    }
}
if (!$libfound && -d "$devkit/libfcgi" && -d "$devkit/include") {
    # devkit
    if (grep { ! -f "$devkit/include/$_" } @dist3 
        or grep { ! -f "$devkit/libfcgi/$_" } @dist2)
    {
        warn "This appears to be a FastCGI devkit distribution, " .
            "but one or more FastCGI library files are missing. \n" .
            "Please check the integrity of the distribution.\n";
        exit -1;
    }

    my $extrarules = join "\n",
        map { $b = $_; $b =~ s/\.c$//; my $s="$devkit/libfcgi/$b.c";
            "$b\$(OBJ_EXT): $s\n\t".
            '$(CCCMD) $(CCCDLFLAGS) -I$(PERL_INC) $(DEFINE) '."$s\n"; }
            @dist2;
    eval 'package MY; sub postamble { $extrarules; }';
    $prefix = $devkit;
}


$sys = $^O eq 'MSWin32' ? 'win32' : 'unix';
push @o, "fcgiapp.o", "os_$sys.o" unless $libfound;
$inc = '-I.' unless $libfound;
$inc .= " -I$prefix/include" if $prefix;

push(@extras, CAPI => 'TRUE')
     if ($] >= 5.005 and $^O eq 'MSWin32'
        and $Config{archname} =~ /-object\b/i);

push(@extras,
    ABSTRACT => 'Fast CGI module',
    AUTHOR   => 'Sven Verdoolaege (skimo@kotnet.org)'
) if ($ExtUtils::MakeMaker::VERSION >= 5.4301);

push @extras, META_MERGE => {
    resources => {
        repository => 'git://git.shadowcat.co.uk/catagits/fcgi2.git',
    },
} if $ExtUtils::MakeMaker::VERSION >= 6.46;

push(@extras,
    MIN_PERL_VERSION => '5.006',
) if $ExtUtils::MakeMaker::VERSION >= 6.48;

# not strictly necessary as everything is in core...
#push(@extras,
#    CONFIGURE_REQUIRES => {
#        ...
#    },
#) if $ExtUtils::MakeMaker::VERSION >= 6.51_03;

if ("$sys" eq "win32") {
    push @libs, ":nosearch -lws2_32";
    push @extras, 'DEFINE' => '-DDLLAPI=__declspec(dllexport)';
}

push @extras,
    'LIBS'    => [ "@libs" ],
    'OBJECT'    => "@o",
    'INC'    => $inc;

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

# Work around bug in previous versions of MakeMaker
WriteMakefile(
    'NAME'        => 'FCGI',
    'VERSION_FROM'    => 'FCGI.pm',
    'dist' => {
        'COMPRESS' => 'gzip -9f',
        'SUFFIX' => 'gz',
        'PREOP' => '$(CP) '.join(' ',
            map {"../$_"} @dist1,
            (map {"libfcgi/$_"} @dist2),
            map {"include/$_"} @dist3).' $(DISTVNAME);'.
                '$(CP) MANIFEST MANIFEST.old;'.
                '$(ECHO) '. join('\\\n',@dist1,@dist2,@dist3) .
                '>> $(DISTVNAME)/MANIFEST',
                  'POSTOP' => 
                '$(MV) MANIFEST.old MANIFEST',
    },
    'clean'        => { FILES => 'config.cache fcgi_config.h fcgi_config.h.in' .
                      ' FCGI.c aclocal.m4 autom4te.cache config.log config.status' .
                      ' FCGI.cfg' },
    'realclean'    => { FILES => 'configure MANIFEST.SKIP.bak MANIFEST.bak Makefile.old' },
    PM            => {'FCGI.pm' => '$(INST_ARCHLIBDIR)/FCGI.pm'},
    PREREQ_PM     => {'XSLoader' => '0'},
    @extras,
);

exit if -f 'fcgi_config.h' or $libfound;

# CPAN and no installed lib found
if ($sys eq "win32") {
    # configure will almost certainly not run on a normal NT install,
    # use the pregenerated configuration file

    print "Using prebuilt fcgi_config.h file for Windows\n";
    unlink("fcgi_config.h");
    my $confdir = $prefix ? "$prefix/include/" : '';
    die $! unless copy("${confdir}fcgi_config_x86.h","fcgi_config.h");
} else {
    print "Running ./configure for you\n";
    print "Please read configure.readme for information on how to run it yourself\n";

    $ENV{'CC'} = $Config{'cc'};
    if ( $^O eq 'android' && !$ENV{'TMPDIR'} ) {
        # See http://stackoverflow.com/a/15417261
        require File::Spec;
        $ENV{'TMPDIR'} = File::Spec->tmpdir();
    }
    system("$Config{sh} configure");
}

